be/src/storage/lake/local_pk_index_manager.cpp

```diff
@@ -86,14 +97,28 @@ void LocalPkIndexManager::gc(UpdateManager* update_manager, DataDir* data_dir, s
         }
         // judge whether tablet should be in the data_dir or not,
         // for data_dir may change if config:storage_path changed.
-        // just remove if not.
-        if (StorageEngine::instance()->get_persistent_index_store(id) != data_dir) {
-            dir_changed_tablet_ids.push_back(id);
-            if (clear_persistent_index(id).ok()) {
-                removed_dir_tablet_ids.push_back(id);
-            } else {
-                gc_fail_because_delete_fail++;
+        // just remove index from old dir if not.
+        auto dir_changed = StorageEngine::instance()->get_persistent_index_store(id) != data_dir;
+        TEST_SYNC_POINT_CALLBACK("LocalPkIndexManager:index_dir_changed:1", &dir_changed);
+        if (dir_changed) {
+            // the shard may be scheduled back to old dir
+            if (!update_manager->try_lock_pk_index_shard(id)) {
+                LOG(WARNING) << "Fail to lock pk index, tablet id: " << id;
+                gc_fail_because_lock_fail++;
+                continue;
             }
+            // double check in the lock
+            dir_changed = StorageEngine::instance()->get_persistent_index_store(id) != data_dir;
+            TEST_SYNC_POINT_CALLBACK("LocalPkIndexManager:index_dir_changed:2", &dir_changed);
+            if (dir_changed) {
+                dir_changed_tablet_ids.push_back(id);
+                if (clear_persistent_index_if_dir_changed(id, data_dir).ok()) {
```

Comment:
here should call `try_remove_primary_index_cache` to remove index cache first to avoid consistency issues

Reply:
done